File: DSLib/Client/UFDS/UFDS.cpp
#include "UFDS.h"
#include <iostream>

int main(int argc, char* argv[])
{
	ds::ufds ufds(10);
	ufds.union_set(1, 8);
	ufds.union_set(8, 9);
	ufds.union_set(8, 3);

	const auto is_same_set = ufds.is_same_set(1, 4);
	std::cout << "is same set (1, 4): " << is_same_set << std::endl;
	return 0;
}

File: DSLib/Lib/UFDS/Ufds.java
package UFDS;

public class Ufds {
    public Ufds(int numNodes) {
        mNumNodes = numNodes;
        mRank = new int[mNumNodes];
        mParentIndex = new int[mNumNodes];

        for (int i = 0; i < mNumNodes; i++) {
            mParentIndex[i] = i;
        }
    }

    public int findSet(int i) {
        if (mParentIndex[i] == i) {
            return i;
        }

        mParentIndex[i] = findSet(mParentIndex[i]);
        return mParentIndex[i];
    }

    public boolean isSameSet(int i, int j) {
        return findSet(i) == findSet(j);
    }

    public void unionSet(int i, int j) {
        if (isSameSet(i, j)) {
            return;
        }

        int iParent = findSet(i);
        int jParent = findSet(j);
        if (mRank[iParent] > mRank[jParent]) {
            mParentIndex[jParent] = iParent;
        } else {
            mParentIndex[iParent] = jParent;
            if (mRank[iParent] == mRank[jParent]) {
                ++mRank[jParent];
            }
        }
    }

    private final int mNumNodes;
    private int[] mParentIndex;
    private int[] mRank;
}
